src/Evolution/Systems/ScalarWave/BoundaryConditions/ConstraintPreservingSphericalRadiation.hpp

```diff
@@ -0,0 +1,267 @@
+#include "Options/Options.hpp"
+#include "Parallel/CharmPupable.hpp"
+#include "PointwiseFunctions/AnalyticData/Tags.hpp"
+#include "PointwiseFunctions/AnalyticSolutions/AnalyticSolution.hpp"
+#include "Time/Tags.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/TMPL.hpp"
+
+/// \cond
+namespace domain::Tags {
+template <size_t Dim, typename Frame>
+struct Coordinates;
+}  // namespace domain::Tags
+/// \endcond
+
+namespace ScalarWave::BoundaryConditions {
+namespace detail {
+/// The type of spherical radiation boundary condition to impose
+enum class ConstraintPreservingSphericalRadiationType {
+  /// Impose \f$(\partial_t + \partial_r)\Psi=0\f$
+  Sommerfeld,
+  /// Impose \f$(\partial_t + \partial_r + r^{-1})\Psi=0\f$
+  FirstOrderBaylissTurkel,
+  /// Imposes a second-order Bayliss-Turkel boundary condition
+  SecondOrderBaylissTurkel
+};
+
+ConstraintPreservingSphericalRadiationType
+convert_constraint_preserving_spherical_radiation_type_from_yaml(
+    const Options::Option& options);
+}  // namespace detail
+
+/*!
+ * \brief Constraint-preserving spherical radiation boundary condition that
+ * seeks to avoid ingoing constraint violations and radiation.
+ *
+ * The constraint-preserving part of the boundary condition imposes the
+ * following condition on the time derivatives of the characteristic fields:
+ *
+ * \f{align*}{
+ *   d_tw^\Psi&\to d_tw^{\Psi}+\lambda_{\Psi}n^i\mathcal{C}_i, \\
+ *   d_tw^0_i&\to d_tw^{0}_i+\lambda_{0}n^jP^k_i\mathcal{C}_{ik},
+ * \f}
+ *
+ * where
+ *
+ * \f{align*}{
+ * P^k{}_i=\delta^k_i-n^kn_i
+ * \f}
+ *
+ * projects a tensor onto the spatial surface to which \f$n_i\f$ is normal, and
+ * \f$d_t w\f$ is the evolved to characteristic field transformation applied to
+ * the time derivatives of the evolved fields. That is,
+ *
+ * \f{align*}{
+ * d_t w^\Psi&=\partial_t \Psi, \\
+ * d_t w_i^0&=(\delta^k_i-n^k n_i)\partial_t \Phi_k, \\
+ * d_t w^{\pm}&=\partial_t\Pi\pm n^k\partial_t\Phi_k - \gamma_2\partial_t\Psi.
+ * \f}
+ *
+ * The constraints are defined as:
+ *
+ * \f{align*}{
+ *  \mathcal{C}_i&=\partial_i\Psi - \Phi_i=0, \\
+ *  \mathcal{C}_{ij}&=\partial_{[i}\Phi_{j]}=0
+ * \f}
+ *
+ * Radiation boundary conditions impose a condition on \f$\Pi\f$ or its time
+ * derivative. We denote the boundary condition value of the time derivative of
+ * \f$\Pi\f$ by \f$\partial_t\Pi^{\mathrm{BC}}\f$. With this, we can impose
+ * boundary conditions on the time derivatives of the evolved variables as
+ * follows:
+ *
+ * \f{align*}{
+ * \partial_{t} \Psi&\to\partial_{t}\Psi +
+ *                    \lambda_\Psi n^i \mathcal{C}_i, \\
+ * \partial_{t}\Pi&\to\partial_{t}\Pi-\left(\partial_t\Pi
+ *                  - \partial_t\Pi^{\mathrm{BC}}\right)
+ *                  +\gamma_2\lambda_\Psi n^i \mathcal{C}_i
+ *                  =\partial_t\Pi^{\mathrm{BC}}
+ *                  +\gamma_2\lambda_\Psi n^i \mathcal{C}_i, \\
+ * \partial_{t}\Phi_i&\to\partial_{t}\Phi_i+
```

Comment:
I'm also not entirely sure what you mean with the `\to` notation here (and for the `d_t w` above). If I understand correctly it means essentially `dt_psi += lambda n^i C_i` on the external boundary. Is the corrected value on the external boundary then combined with the interior value of dt_psi through a "ghost" element of sorts? No need to explain details here, I'm just wondering if a clearer notation would be something like `\partial_t \Psi_\mathrm{exterior} = \partial_t \Psi + \lambda_\Psi n^i C_i`, since I know much better what `=` means than what `\to` means 😄

Reply:
There isn't a ghost element here. You could probably reformulate these as continuous Galerkin-looking condition with a ghost element, but that would probably be even more confusing. There isn't a `\dt \Psi_ext`, so I don't know what your equation means. Bjorhus boundary conditions only deal with the interior information (I guess you could impose Dirichlet BCs via Bjorhus, but I haven't ever seen that done in practice). If you would do the CG approach, then `dt Psi_ext=dt Psi_int + 2 lambda n^i C_i` note the factor of 2, which is important.